backend/oasst_backend/tree_manager.py

```diff
@@ -0,0 +1,760 @@
+        protocol_schema.TextLabel.helpful,
+        protocol_schema.TextLabel.spam,
+    ]
+    mandatory_labels_prompter_reply: Optional[list[protocol_schema.TextLabel]] = [
+        protocol_schema.TextLabel.helpful,
+        protocol_schema.TextLabel.spam,
+    ]
+
+
+class TaskType(Enum):
+    NONE = -1
+    RANKING = 0
+    LABEL_REPLY = 1
+    REPLY = 2
+    LABEL_PROMPT = 3
+    PROMPT = 4
+
+
+class ActiveTreeSizeRow(pydantic.BaseModel):
+    message_tree_id: UUID
+    tree_size: int
+    goal_tree_size: int
+
+    @property
+    def remaining_messages(self) -> int:
+        return max(0, self.goal_tree_size - self.tree_size)
+
+    class Config:
+        orm_mode = True
+
+
+class ExtendibleParentRow(pydantic.BaseModel):
+    parent_id: UUID
+    depth: int
+    message_tree_id: UUID
+    active_children_count: int
+
+    class Config:
+        orm_mode = True
+
+
+class IncompleteRankingsRow(pydantic.BaseModel):
+    parent_id: UUID
+    children_count: int
+    child_min_ranking_count: int
+
+    class Config:
+        orm_mode = True
+
+
+class TreeManager:
+    def __init__(self, db: Session, prompt_repository: PromptRepository, configuration: TreeManagerConfiguration):
+        self.db = db
+        self.cfg = configuration
+        self.pr = prompt_repository
+
+    def _task_selection(
+        self,
+        num_ranking_tasks: int,
+        num_replies_need_review: int,
+        num_prompts_need_review: int,
+        num_missing_prompts: int,
+        num_missing_replies: int,
+    ) -> TaskType:
+        """
+        Determines which task to hand out to human worker.
+        The task type is drawn with relative weight (e.g. ranking has highest priority)
+        depending on what is possible with the current message trees in the database.
+        """
+
+        logger.debug(
+            f"TreeManager._task_selection({num_ranking_tasks=}, {num_replies_need_review=}, "
+            f"{num_prompts_need_review=}, {num_missing_prompts=}, {num_missing_replies=})"
+        )
+
+        task_weights = [0] * 5
+
+        if num_ranking_tasks > 0:
+            task_weights[TaskType.RANKING.value] = 10
+
+        if num_replies_need_review > 0:
+            task_weights[TaskType.LABEL_REPLY.value] = 5
+
+        if num_prompts_need_review > 0:
+            task_weights[TaskType.LABEL_PROMPT.value] = 5
+
+        if num_missing_replies > 0:
+            task_weights[TaskType.REPLY.value] = 2
+
+        if num_missing_prompts > 0:
+            task_weights[TaskType.PROMPT.value] = 1
+
+        task_weights = np.array(task_weights)
+        weight_sum = task_weights.sum()
+        if weight_sum < 1e-8:
+            task_type = TaskType.NONE
+        else:
+            task_weights = task_weights / weight_sum
+            task_type = TaskType(np.random.choice(a=len(task_weights), p=task_weights))
+
+        logger.debug(f"Selected task type: {task_type}")
+        return TaskType(task_type)
```

Comment:
isn't `task_type` already a `TaskType`?

Reply:
yes, I think added the cast to the np.random.choice and forgot to remove it there.